File: mockserver/mockserver-core/src/main/java/org/mockserver/model/HttpLlmResponse.java
package org.mockserver.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.mockserver.matchers.LlmConversationMatcher;

import java.util.Objects;

public class HttpLlmResponse extends Action<HttpLlmResponse> {
    private int hashCode;
    private Provider provider;
    private String model;
    private Completion completion;
    private EmbeddingResponse embedding;
    private RerankResponse rerank;
    private ConversationPredicates conversationPredicates;
    private LlmChaosProfile chaos;
    // volatile: lazily reconstructed (see getConversationMatcher) on the concurrent
    // match path from a shared expectation object. The value is effectively immutable
    // once set, so volatile guarantees safe publication under the Java Memory Model — a
    // thread that observes a non-null reference also observes a fully-constructed matcher.
    @JsonIgnore
    private transient volatile LlmConversationMatcher conversationMatcher;

    public static HttpLlmResponse llmResponse() {
        return new HttpLlmResponse();
    }

    public HttpLlmResponse withProvider(Provider provider) {
        this.provider = provider;
        this.hashCode = 0;
        return this;
    }

    public Provider getProvider() {
        return provider;
    }

    public HttpLlmResponse withModel(String model) {
        this.model = model;
        this.hashCode = 0;
        return this;
    }

    public String getModel() {
        return model;
    }

    public HttpLlmResponse withCompletion(Completion completion) {
        this.completion = completion;
        this.hashCode = 0;
        return this;
    }

    public Completion getCompletion() {
        return completion;
    }

    public HttpLlmResponse withEmbedding(EmbeddingResponse embedding) {
        this.embedding = embedding;
        this.hashCode = 0;
        return this;
    }

    public EmbeddingResponse getEmbedding() {
        return embedding;
    }

    public HttpLlmResponse withRerank(RerankResponse rerank) {
        this.rerank = rerank;
        this.hashCode = 0;
        return this;
    }

    public RerankResponse getRerank() {
        return rerank;
    }

    public HttpLlmResponse withConversationPredicates(ConversationPredicates conversationPredicates) {
        this.conversationPredicates = conversationPredicates;
        this.hashCode = 0;
        this.conversationMatcher = null; // invalidate cached matcher
        return this;
    }

    public ConversationPredicates getConversationPredicates() {
        return conversationPredicates;
    }

    /**
     * Optional fault/chaos profile applied when returning this response —
     * probabilistic errors, mid-stream truncation, malformed SSE.
     */
    public HttpLlmResponse withChaos(LlmChaosProfile chaos) {
        this.chaos = chaos;
        this.hashCode = 0;
        return this;
    }

    public LlmChaosProfile getChaos() {
        return chaos;
    }

    public HttpLlmResponse withConversationMatcher(LlmConversationMatcher conversationMatcher) {
        this.conversationMatcher = conversationMatcher;
        return this;
    }

    /**
     * Returns the conversation matcher for evaluation. If not explicitly set,
     * lazily reconstructs it from {@link #conversationPredicates} (which survives
     * JSON round-tripping). Returns null if neither is set.
     */
    @JsonIgnore
    public LlmConversationMatcher getConversationMatcher() {
        if (conversationMatcher == null && conversationPredicates != null && conversationPredicates.hasAnyPredicate()) {
            LlmConversationMatcher matcher = new LlmConversationMatcher();
            matcher.withProvider(provider);
            if (conversationPredicates.getTurnIndex() != null) {
                matcher.withTurnIndex(conversationPredicates.getTurnIndex());
            }
            if (conversationPredicates.getLatestMessageContains() != null) {
                matcher.withLatestMessageContains(conversationPredicates.getLatestMessageContains());
            }
            if (conversationPredicates.getLatestMessageMatches() != null) {
                matcher.withLatestMessageMatches(conversationPredicates.getLatestMessageMatches());
            }
            if (conversationPredicates.getLatestMessageRole() != null) {
                matcher.withLatestMessageRole(conversationPredicates.getLatestMessageRole());
            }
            if (conversationPredicates.getContainsToolResultFor() != null) {
                matcher.withContainsToolResultFor(conversationPredicates.getContainsToolResultFor());
            }
            if (conversationPredicates.getSemanticMatchAgainst() != null) {
                matcher.withSemanticMatchAgainst(conversationPredicates.getSemanticMatchAgainst());
            }
            if (conversationPredicates.getNormalization() != null) {
                matcher.withNormalization(conversationPredicates.getNormalization());
            }
            conversationMatcher = matcher;
        }
        return conversationMatcher;
    }

    @Override
    @JsonIgnore
    public Type getType() {
        return Type.LLM_RESPONSE;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        if (hashCode() != o.hashCode()) {
            return false;
        }
        if (!super.equals(o)) {
            return false;
        }
        HttpLlmResponse that = (HttpLlmResponse) o;
        return Objects.equals(provider, that.provider) &&
            Objects.equals(model, that.model) &&
            Objects.equals(completion, that.completion) &&
            Objects.equals(embedding, that.embedding) &&
            Objects.equals(rerank, that.rerank) &&
            Objects.equals(conversationPredicates, that.conversationPredicates) &&
            Objects.equals(chaos, that.chaos);
    }

    @Override
    public int hashCode() {
        if (hashCode == 0) {
            hashCode = Objects.hash(super.hashCode(), provider, model, completion, embedding, rerank, conversationPredicates, chaos);
        }
        return hashCode;
    }
}

File: mockserver-client-php/src/Llm/HttpLlmResponse.php
<?php

declare(strict_types=1);

namespace MockServer\Llm;

use MockServer\Delay;

/**
 * The {@code httpLlmResponse} action payload of an LLM expectation
 * (mirrors org.mockserver.model.HttpLlmResponse).
 *
 * Sibling of {@code httpRequest}, {@code httpResponse}, {@code scenarioName},
 * {@code scenarioState} and {@code newScenarioState} on an expectation.
 */
class HttpLlmResponse implements \JsonSerializable
{
    private ?string $provider = null;
    private ?string $model = null;
    private ?Completion $completion = null;
    private ?EmbeddingResponse $embedding = null;
    private ?ConversationPredicates $conversationPredicates = null;
    /** @var array<string, mixed>|null */
    private ?array $chaos = null;
    private ?Delay $delay = null;

    /**
     * Static factory mirroring {@code HttpLlmResponse.llmResponse()}.
     */
    public static function llmResponse(): self
    {
        return new self();
    }

    public function withProvider(string $provider): self
    {
        $this->provider = $provider;
        return $this;
    }

    public function withModel(string $model): self
    {
        $this->model = $model;
        return $this;
    }

    public function withCompletion(Completion $completion): self
    {
        $this->completion = $completion;
        $this->embedding = null;
        return $this;
    }

    public function withEmbedding(EmbeddingResponse $embedding): self
    {
        $this->embedding = $embedding;
        $this->completion = null;
        return $this;
    }

    public function withConversationPredicates(ConversationPredicates $conversationPredicates): self
    {
        $this->conversationPredicates = $conversationPredicates;
        return $this;
    }

    /**
     * @param array<string, mixed> $chaos
     */
    public function withChaos(array $chaos): self
    {
        $this->chaos = $chaos;
        return $this;
    }

    public function withDelay(Delay $delay): self
    {
        $this->delay = $delay;
        return $this;
    }

    /**
     * @return array<string, mixed>
     */
    public function jsonSerialize(): array
    {
        return $this->toArray();
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $data = [];
        if ($this->provider !== null) {
            $data['provider'] = $this->provider;
        }
        if ($this->model !== null) {
            $data['model'] = $this->model;
        }
        if ($this->completion !== null) {
            $data['completion'] = $this->completion->toArray();
        }
        if ($this->embedding !== null) {
            $data['embedding'] = $this->embedding->toArray();
        }
        if ($this->conversationPredicates !== null) {
            $data['conversationPredicates'] = $this->conversationPredicates->toArray();
        }
        if ($this->chaos !== null) {
            $data['chaos'] = $this->chaos;
        }
        if ($this->delay !== null) {
            $data['delay'] = $this->delay->toArray();
        }
        return $data;
    }
}
